crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -879,12 +890,17 @@ where
                 walk_expr(self, expr);
             }
             ast::Expr::Named(node) => {
-                debug_assert!(self.current_assignment.is_none());
-                // TODO walrus in comprehensions is implicitly nonlocal
-                self.visit_expr(&node.value);
-                self.current_assignment = Some(node.into());
-                self.visit_expr(&node.target);
-                self.current_assignment = None;
+                if self.current_assignment.is_some() {
+                    // This can happen if we have something like x = y := 2
+                    // which is invalid syntax but still is provided in the AST
+                    tracing::warn!("Current assignment is unexpectedly set");
```

Comment:
We limit the use of `tracing::warn` for messages that need the user's attention and they need to address (e.g. when setting up file watching for a directory failed). Here, there's not much a user can do about this and invalid syntax is very common in an LSP. Logging a warning would be very noisy for users. We should use either the `debug` or possibly even `trace` level instead.